### PR TITLE
Ignore nested events in GTK shell

### LINKED ISSUE #571: The SHOW_OPEN_PANEL command panics on gtk
[This](https://github.com/xi-editor/druid/blob/master/druid-shell/src/platform/gtk/window.rs#L318) is where the panic happens. I'm not sure what the real root cause is. That code is new from #552, but [this comment](https://github.com/xi-editor/druid/blob/master/druid/src/win_handler.rs#L520) looks suspicious.

backtrace:
```
thread '<unnamed>' panicked at 'already borrowed: BorrowMutError', src/libcore/result.rs:1165:5
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:77
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:61
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1028
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1412
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:65
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:50
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:188
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:205
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:464
  11: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:373
  12: rust_begin_unwind
             at src/libstd/panicking.rs:302
  13: core::panicking::panic_fmt
             at src/libcore/panicking.rs:139
  14: core::result::unwrap_failed
             at src/libcore/result.rs:1165
  15: core::result::Result<T,E>::expect
             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libcore/result.rs:960
  16: core::cell::RefCell<T>::borrow_mut
             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libcore/cell.rs:869
  17: druid_shell::platform::gtk::window::WindowBuilder::build::{{closure}}
             at /home/dan/.cargo/git/checkouts/druid-1a6b6c4d20db75fe/a0de2a1/druid-shell/src/platform/gtk/window.rs:316
  18: <O as gtk::auto::widget::WidgetExt>::connect_leave_notify_event::leave_notify_event_trampoline
             at /home/dan/.cargo/registry/src/github.com-1ecc6299db9ec823/gtk-0.8.1/src/auto/widget.rs:4154
  19: <unknown>
  20: <unknown>
  21: g_signal_emit_valist
  22: g_signal_emit
  23: <unknown>
  24: <unknown>
  25: <unknown>
  26: <unknown>
  27: <unknown>
  28: <unknown>
  29: <unknown>
  30: <unknown>
  31: <unknown>
  32: <unknown>
  33: g_closure_invoke
  34: <unknown>
  35: g_signal_emit_valist
  36: g_signal_emit
  37: gtk_widget_show
  38: gtk_native_dialog_show
  39: gtk_native_dialog_run
  40: <O as gtk::auto::native_dialog::NativeDialogExt>::run
             at /home/dan/.cargo/registry/src/github.com-1ecc6299db9ec823/gtk-0.8.1/src/auto/native_dialog.rs:152
  41: druid_shell::platform::gtk::dialog::get_file_dialog_path
             at /home/dan/.cargo/git/checkouts/druid-1a6b6c4d20db75fe/a0de2a1/druid-shell/src/platform/gtk/dialog.rs:44
  42: druid_shell::platform::gtk::window::WindowHandle::file_dialog
             at /home/dan/.cargo/git/checkouts/druid-1a6b6c4d20db75fe/a0de2a1/druid-shell/src/platform/gtk/window.rs:584
  43: druid_shell::platform::gtk::window::WindowHandle::open_file_sync
             at /home/dan/.cargo/git/checkouts/druid-1a6b6c4d20db75fe/a0de2a1/druid-shell/src/platform/gtk/window.rs:480
  44: druid_shell::window::WindowHandle::open_file_sync
             at /home/dan/.cargo/git/checkouts/druid-1a6b6c4d20db75fe/a0de2a1/druid-shell/src/window.rs:163
  45: druid::win_handler::AppState<T>::show_open_panel::{{closure}}
             at /home/dan/.cargo/git/checkouts/druid-1a6b6c4d20db75fe/a0de2a1/druid/src/win_handler.rs:530
  46: core::option::Option<T>::and_then
             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libcore/option.rs:655
  47: druid::win_handler::AppState<T>::show_open_panel
             at /home/dan/.cargo/git/checkouts/druid-1a6b6c4d20db75fe/a0de2a1/druid/src/win_handler.rs:530
  48: druid::win_handler::AppState<T>::handle_cmd
             at /home/dan/.cargo/git/checkouts/druid-1a6b6c4d20db75fe/a0de2a1/druid/src/win_handler.rs:506
  49: druid::win_handler::AppState<T>::process_commands
             at /home/dan/.cargo/git/checkouts/druid-1a6b6c4d20db75fe/a0de2a1/druid/src/win_handler.rs:457
  50: druid::win_handler::AppState<T>::do_window_event
             at /home/dan/.cargo/git/checkouts/druid-1a6b6c4d20db75fe/a0de2a1/druid/src/win_handler.rs:433
  51: <druid::win_handler::DruidHandler<T> as druid_shell::window::WinHandler>::mouse_up
             at /home/dan/.cargo/git/checkouts/druid-1a6b6c4d20db75fe/a0de2a1/druid/src/win_handler.rs:634
  52: druid_shell::platform::gtk::window::WindowBuilder::build::{{closure}}
             at /home/dan/.cargo/git/checkouts/druid-1a6b6c4d20db75fe/a0de2a1/druid-shell/src/platform/gtk/window.rs:272
  53: <O as gtk::auto::widget::WidgetExt>::connect_button_release_event::button_release_event_trampoline
             at /home/dan/.cargo/registry/src/github.com-1ecc6299db9ec823/gtk-0.8.1/src/auto/widget.rs:3096
  54: <unknown>
  55: g_closure_invoke
  56: <unknown>
  57: g_signal_emit_valist
  58: g_signal_emit
  59: <unknown>
  60: <unknown>
  61: gtk_main_do_event
  62: <unknown>
  63: <unknown>
  64: g_main_context_dispatch
  65: <unknown>
  66: g_main_context_iteration
  67: g_application_run
  68: <O as gio::application::ApplicationExtManual>::run
             at /home/dan/.cargo/registry/src/github.com-1ecc6299db9ec823/gio-0.8.1/src/application.rs:23
  69: druid_shell::platform::gtk::application::Application::run::{{closure}}
             at /home/dan/.cargo/git/checkouts/druid-1a6b6c4d20db75fe/a0de2a1/druid-shell/src/platform/gtk/application.rs:70
  70: std::thread::local::LocalKey<T>::try_with
             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libstd/thread/local.rs:262
  71: std::thread::local::LocalKey<T>::with
             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libstd/thread/local.rs:239
  72: druid_shell::platform::gtk::application::Application::run
             at /home/dan/.cargo/git/checkouts/druid-1a6b6c4d20db75fe/a0de2a1/druid-shell/src/platform/gtk/application.rs:69
  73: druid_shell::application::Application::run
             at /home/dan/.cargo/git/checkouts/druid-1a6b6c4d20db75fe/a0de2a1/druid-shell/src/application.rs:53
  74: druid::app::AppLauncher<T>::launch
             at /home/dan/.cargo/git/checkouts/druid-1a6b6c4d20db75fe/a0de2a1/druid/src/app.rs:124
  75: music_player::launch_ui
             at src/lib.rs:142
  76: music_player::main::{{closure}}
             at src/main.rs:13

```

### LINKED ISSUE #735: context menu pop up causes crash
Right click on the multiwin example gives me this:

```
cargo run --example multiwin
   Compiling druid v0.5.0 (/home/kaiyin/rustProjects/druid/druid)
    Finished dev [unoptimized + debuginfo] target(s) in 3.80s
     Running `/home/kaiyin/rustProjects/druid/target/debug/examples/multiwin`
WARN  [druid::menu] MenuDesc::platform_default is not implemented for this platform.
DEBUG [druid::localization] available locales [en-US, de-DE, fr-CA], current en-US
DEBUG [druid::localization] resolved: [en-US]
WARN  [druid::menu] MenuDesc::platform_default is not implemented for this platform.
INFO  [multiwin] Window added, id: WindowId(1)
INFO  [druid_shell::platform::gtk::application] gtk: Activated application
thread 'main' panicked at 'already borrowed: BorrowMutError', /rustc/f509b26a7730d721ef87423a72b3fdf8724b4afa/src/libcore/cell.rs:878:9
stack backtrace:
   0:     0x560fa3665c64 - <unknown>
   1:     0x560fa36852bc - <unknown>
   2:     0x560fa3663a93 - <unknown>
   3:     0x560fa3668295 - <unknown>
   4:     0x560fa3667fe2 - <unknown>
   5:     0x560fa36688f2 - <unknown>
   6:     0x560fa36684db - <unknown>
   7:     0x560fa3683c81 - <unknown>
   8:     0x560fa3683aa3 - <unknown>
   9:     0x560fa3651550 - <unknown>
  10:     0x7fc689bcb7ff - <unknown>
  11:     0x7fc68931b936 - <unknown>
  12:     0x7fc689337dcb - g_signal_emit_valist
  13:     0x7fc6893389b3 - g_signal_emit
  14:     0x7fc689b75e53 - <unknown>
  15:     0x7fc689b760ec - <unknown>
  16:     0x7fc689b7ca27 - <unknown>
  17:     0x7fc689a31b88 - <unknown>
  18:     0x7fc689a31f49 - <unknown>
  19:     0x7fc689915639 - <unknown>
  20:     0x7fc689a31e60 - <unknown>
  21:     0x7fc689b8bf19 - <unknown>
  22:     0x7fc689a31e60 - <unknown>
  23:     0x7fc689a32008 - <unknown>
  24:     0x7fc689a3b233 - <unknown>
  25:     0x7fc689a3b755 - gtk_menu_popup
  26:     0x560fa364d8f7 - <unknown>
  27:     0x560fa364b250 - <unknown>
  28:     0x560fa361f6ba - <unknown>
  29:     0x560fa361ca4a - <unknown>
  30:     0x560fa361dafa - <unknown>
  31:     0x560fa361e081 - <unknown>
  32:     0x560fa361df0c - <unknown>
  33:     0x560fa361e9cf - <unknown>
  34:     0x560fa36511c0 - <unknown>
  35:     0x7fc689bcb70b - <unknown>
  36:     0x7fc68931b6e2 - g_closure_invoke
  37:     0x7fc68932f544 - <unknown>
  38:     0x7fc68933799e - g_signal_emit_valist
  39:     0x7fc6893389b3 - g_signal_emit
  40:     0x7fc689b75e53 - <unknown>
  41:     0x7fc689a319d8 - <unknown>
  42:     0x7fc689a33c2b - gtk_main_do_event
  43:     0x7fc68971df19 - <unknown>
  44:     0x7fc689750ff6 - <unknown>
  45:     0x7fc68923084d - g_main_context_dispatch
  46:     0x7fc689230ad0 - <unknown>
  47:     0x7fc689230b73 - g_main_context_iteration
  48:     0x7fc689443b05 - g_application_run
  49:     0x560fa364e96c - <unknown>
  50:     0x560fa364af76 - <unknown>
  51:     0x560fa364a996 - <unknown>
  52:     0x560fa3615471 - <unknown>
  53:     0x560fa3619853 - <unknown>
  54:     0x560fa3614340 - <unknown>
  55:     0x560fa3668dbf - <unknown>
  56:     0x560fa361a39c - <unknown>
  57:     0x7fc688fd01e3 - __libc_start_main
  58:     0x560fa361018e - <unknown>
  59:                0x0 - <unknown>
```

### PR DESCRIPTION
**Pull request by Finnerale**
Now the GTK shell uses `Mutex::try_borrow_mut` and thus ignores nested events. 

This Closes #735 and Closes #571

### CODE REVIEW

**Comment by xStrom** [comment]
What happens to the event when it's ignored? Just dropped? I think a `log_dropped_msg` like the one in the Windows implementation would be a good idea here.

File: druid-shell/src/platform/gtk/window.rs [hunk 1]
```diff
@@ -421,8 +438,11 @@ impl WindowBuilder {
 
                 *(state.current_keyval.borrow_mut()) = None;
 
-                let key_event = make_key_event(key, false);
-                state.handler.borrow_mut().key_up(key_event);
+                if let Ok(mut handler) = state.handler.try_borrow_mut() {
+                    handler.key_down(make_key_event(key, false));
```
Comment (xStrom) [inline druid-shell/src/platform/gtk/window.rs:442]: Copy-paste mistake? Shouldn't this still be `key_up`?

File: druid-shell/src/platform/gtk/window.rs [hunk 2]
```diff
@@ -479,7 +499,7 @@ impl WindowHandle {
     /// Bring this window to the front of the window stack and give it focus.
     pub fn bring_to_front_and_focus(&self) {
         //FIXME: implementation goes here
-        log::warn!("bring_to_front_and_focus not yet implemented for gtk");
+        warn!("bring_to_front_and_focus not yet implemented for gtk");
```
Comment (xStrom) [inline druid-shell/src/platform/gtk/window.rs:502]: I think @cmyr wants us to not use imported functions directly. That is `log::warn!` is better than `warn!`. It's from the [Rust style guide](https://doc.rust-lang.org/1.0.0/style/style/imports.html#prefer-fully-importing-types/traits-while-module-qualifying-functions).

Comment (Finnerale) [reply druid-shell/src/platform/gtk/window.rs]: I prefer that too, was just done differently in the Windows version I used as reference.
Wonder if there are lints for such things :thinking:

Comment (xStrom) [reply druid-shell/src/platform/gtk/window.rs]: Yeah the windows platform code has some of the oldest code in druid and so follows style the least.

I've been also thinking about looking into custom clippy/rustfmt options to enforce some of our style rules.

**Comment by Finnerale** [comment]
Well, I still do not truly understand how how druid-shell works but from what I've seen I doubt this can be truly solved without restructuring druid-shell. Also ignoring nested events does not really hurt the user experience as they are rare and do not matter in the cases where they occur.

File: druid-shell/src/platform/gtk/window.rs (hunk 2 above)
Comment (cmyr) [reply druid-shell/src/platform/gtk/window.rs]: yea, this is a nit but I like macros and functions to be module-qualified. Would gladly accept a PR that standardizes on this :)

**Comment by cmyr** [comment]
yes, i think at least being defensive is the right decision for the time being.

**Pull request merged**